decidim-core/app/controllers/decidim/locales_controller.rb

```diff
@@ -0,0 +1,19 @@
+# frozen_string_literal: true
+require_dependency "decidim/application_controller"
+
+module Decidim
+  # A controller to allow users switching their locale.
+  class LocalesController < ApplicationController
+    def create
+      if current_user && params["locale"] && available_locales.include?(params["locale"])
+        current_user.update_attribute(:locale, params["locale"])
+      end
+
+      uri = URI(params["back"] || "/")
+      params = URI.decode_www_form(uri.query || "") << ["locale", current_locale]
```

Comment:
This whole block looks a bit convoluted and it's mainly caused because of the URL needing to carry that `locale` parameter around. I'm not sure I'm cool with that. 

If someone sends a link to a resource, I don't care about the language if the *content* is the same. But I would care if the interface language changed.

I'm -1 in the whole `locale` param. For me, an URL should point to the same *resource*, not the same *exact content*. If a polish-speaking user shares some content and an english-speaking user opens it, it would be way more valuable to me to get it in my own language.